README.md
# varnishd cookbook

A [Chef](http://docs.opscode.com/#) [cookbook](http://docs.opscode.com/essentials_cookbooks.html) for installing and configuring [Varnish](https://www.varnish-cache.org/), with some specific design constraints.

* Varnish and [VMODs](https://www.varnish-cache.org/vmods) are installed from **source**.
* ***Every*** tunable parameter is exposed in the [attributes](http://docs.opscode.com/chef_overview_attributes.html).
* Services are managed by [Upstart](http://upstart.ubuntu.com/) to provide straight-forward resource limits, respawing, and control.
* Carefully triggered and idempotent build/dependency chains to insure Chef runs cleanly in the event of errors.

## Requirements

### Platforms

Using [test kitchen](https://github.com/test-kitchen/test-kitchen), this cookbook is tested against the following platforms.

- Ubuntu 12.04, Ubuntu 12.10, Ubuntu 13.04

It'd be reasonable to assume that Ubuntu versions more recent than 13.04 will work as well as Debian and older Ubuntu distributions that have Upstart configured.

### Cookbooks

- apt
- build-essential
- git

## Usage

Add `varnishd::default` to your node's [run-list](http://docs.opscode.com/essentials_node_object_run_lists.html). If you do not want to setup the `varnishd` daemon as a service, you may include `varnishd::build` and `varnishd::vmods` seperately instead.

## Attributes

It's important to note that the `:version`, `:url`, and `:checksum` attributes must always be modified together and are in no way automatically generated.

key | description | default
--- | --- | ---
[:varnishd][:version] | Complete Varnish version you are installing | '3.0.5'
[:varnishd][:url] | URL for an source archive of that version | 'http://repo.varnish-cache.org/source/varnish-3.0.5.tar.gz'
[:varnishd][:checksum] | Checksum for that archive | '302fd6afc7…'
[:varnishd][:vcl_source] | Template to use to render the VCL | 'default.vcl.erb'
[:varnishd][:vcl_cookbook] | Cookbook in which to find that template | 'varnishd'
[:varnishd][:storage] | Storage type and options | 'malloc,128M'
[:varnishd][:pid] | Write the daemon's PID to file | '/var/run/varnishd.pid'
[:varnishd][:limit][:nofile] | File limit | '131072'
[:varnishd][:limit][:memlock] | Memory limit | '82000'

The mechanism for specifying a VCL template in a wrapping cookbook is the same as in the [varnish cookbook](https://github.com/opscode-cookbooks/varnish).

### VMODs

VMODs to build and install can be specified as part of the `[:varnishd][:vmods]` attribute. For example, to include [boltsort](https://github.com/vimeo/libvmod-boltsort).

```ruby
node.set[:varnishd][:vmods][:boltsort][:repository] = 'https://github.com/vimeo/libvmod-boltsort.git'
node.set[:varnishd][:vmods][:boltsort][:reference] = 'master'
```

The `:reference` is optional, and if it is missing `master` will be used by default.

### Varnish Runtime Options

A key feature of this cookbook is providing access to *every* Varnish runtime option as described in [the documentation](https://www.varnish-cache.org/docs/3.0/reference/varnishd.html). These options are listed out in the default [attribute file](http://docs.opscode.com/essentials_cookbook_attribute_files.html) like so.

```ruby
default[:varnishd][:runtime][:http_max_hdr] = nil #'64'
```

For most values the default is `nil` and the value that Varnish will use when the value is not provided is listed in end of line comment. Any value not set to nil will be passed to `varnishd` when run using the `-p` flag.

## Contributing

Please do! We'll follow the [git-flow](http://nvie.com/posts/a-successful-git-branching-model/) method, so please submit pull requests as feature branches.

## Contributors

* **[@buth](https://github.com/buth)**
* **[@abstrctn](https://github.com/abstrctn)**


CHANGELOG.md
# varnishd cookbook CHANGELOG
This file is used to list changes made in each version of the varnishd cookbook.

###v0.1.3
* updated package requirements for build

###v0.1.2
* PID file attribute (**[@abstrctn](https://github.com/abstrctn)**)
* metadata update


recipes/build.rb
#
# Cookbook Name:: varnishd
# Recipe:: build
#
# Copyright (C) 2014 Eric Buth
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy
# of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.
#

include_recipe 'apt'
include_recipe 'build-essential'

%w[
  autoconf
  autotools-dev
  groff-base
  libedit-dev
  libncurses-dev
  libpcre3-dev
  libtool
  pkg-config
  python-docutils
  xsltproc
].each do |name|
  package name
end

%w[
  /usr/local/src
  /usr/local/etc/varnish
].each do |path|
  directory path do
    recursive true
  end
end

execute 'varnish-secret' do
  command 'uuidgen > /usr/local/etc/varnish/secret && chmod 0600 /usr/local/etc/varnish/secret'
  creates '/usr/local/etc/varnish/secret'
end

remote_file "/usr/local/src/varnish-#{node[:varnishd][:version]}.tar.gz" do
  source node[:varnishd][:url]
  checksum node[:varnishd][:checksum]
end

execute 'varnish-unpack' do
  command "tar -zxf varnish-#{node[:varnishd][:version]}.tar.gz"
  cwd '/usr/local/src'
  creates "/usr/local/src/varnish-#{node[:varnishd][:version]}"
end  

execute 'varnish-build' do
  command './autogen.sh && ./configure && make'
  cwd "/usr/local/src/varnish-#{node[:varnishd][:version]}"
  creates "/usr/local/src/varnish-#{node[:varnishd][:version]}/bin/varnishd/varnishd"
end

link '/usr/local/src/varnish' do
  to "/usr/local/src/varnish-#{node[:varnishd][:version]}"
end

execute 'varnish-install' do
  command 'make install'
  cwd '/usr/local/src/varnish'
  not_if "/usr/local/sbin/varnishd -V 2>&1 | grep -q '\\bvarnish-#{node[:varnishd][:version]}\\b'"
  notifies :restart, 'service[varnish]', :delayed
end


recipes/vmods.rb
#
# Cookbook Name:: varnishd
# Recipe:: vmods
#
# Copyright (C) 2014 Eric Buth
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy
# of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.
#

include_recipe 'varnishd::build'
include_recipe 'git'

directory node[:varnishd][:runtime][:vmod_dir] do
  recursive true
end

node[:varnishd][:vmods].each_pair do |vmod, attributes|

  git "/usr/local/src/libvmod-#{vmod}" do
    repository attributes[:repository]
    reference attributes[:reference].to_s == '' ? 'master' : attributes[:reference]
    action :sync
  end

  execute "varnish-vmod-build-#{vmod}" do
    cwd "/usr/local/src/libvmod-#{vmod}"
    command './autogen.sh && ./configure && make && make check && make install'
    environment(
      'VARNISHSRC' => '/usr/local/src/varnish',
      'VMODDIR' => node[:varnishd][:runtime][:vmod_dir]
    )
    notifies :restart, 'service[varnish]', :delayed
    only_if do
      !::File.exists?("/usr/local/src/libvmod-#{vmod}/configure") ||
      ::File.mtime("/usr/local/src/libvmod-#{vmod}/configure") < ::File.mtime("/usr/local/src/libvmod-#{vmod}/.git") ||
      ::File.mtime("/usr/local/src/libvmod-#{vmod}/configure") < ::File.mtime('/usr/local/src/varnish')
    end
  end
end


attributes/default.rb
default[:varnishd][:version] = '3.0.5'
default[:varnishd][:url] = 'http://repo.varnish-cache.org/source/varnish-3.0.5.tar.gz'
default[:varnishd][:checksum] = '302fd6afc771524ca3912f5d945ab197a55762385c012b2054df7d86bf7ae2b7'

# VCL
default[:varnishd][:vcl_source] = 'default.vcl.erb'
default[:varnishd][:vcl_cookbook] = 'varnishd'

# Storage Types
# https://www.varnish-cache.org/docs/3.0/reference/varnishd.html#storage-types
default[:varnishd][:storage] = 'malloc,128M'

# PID file location
default[:varnishd][:pid] = '/var/run/varnishd.pid'

# System Limits
default[:varnishd][:limit][:nofile] = '131072'
default[:varnishd][:limit][:memlock] = '82000'

# VMODs
default[:varnishd][:vmods] = {}

# Run-Time Parameters
# https://www.varnish-cache.org/docs/3.0/reference/varnishd.html#run-time-parameters
default[:varnishd][:runtime][:acceptor_sleep_decay] = nil #'0.900'
default[:varnishd][:runtime][:acceptor_sleep_incr] = nil #'0.001'
default[:varnishd][:runtime][:acceptor_sleep_max] = nil #'0.050'
default[:varnishd][:runtime][:auto_restart] = nil #'on'
default[:varnishd][:runtime][:ban_dups] = nil #'on'
default[:varnishd][:runtime][:ban_lurker_sleep] = nil #'0.01'
default[:varnishd][:runtime][:between_bytes_timeout] = nil #'60'
default[:varnishd][:runtime][:cc_command] = nil #'exec gcc -std=gnu99 -pthread -fpic -shared -Wl,-x -o %o %s'
default[:varnishd][:runtime][:cli_buffer] = nil #'8192'
default[:varnishd][:runtime][:cli_timeout] = nil #'10'
default[:varnishd][:runtime][:clock_skew] = nil #'10'
default[:varnishd][:runtime][:connect_timeout] = nil #'0.7'
default[:varnishd][:runtime][:critbit_cooloff] = nil #'180.0'
default[:varnishd][:runtime][:default_grace] = nil #'10'
default[:varnishd][:runtime][:default_keep] = nil #'0'
default[:varnishd][:runtime][:default_ttl] = nil #'120'
default[:varnishd][:runtime][:diag_bitmap] = nil #'0'
default[:varnishd][:runtime][:esi_syntax] = nil #'0'
default[:varnishd][:runtime][:expiry_sleep] = nil #'1'
default[:varnishd][:runtime][:fetch_chunksize] = nil #'128'
default[:varnishd][:runtime][:fetch_maxchunksize] = nil #'262144'
default[:varnishd][:runtime][:first_byte_timeout] = nil #'60'
default[:varnishd][:runtime][:group] = 'varnishd'
default[:varnishd][:runtime][:gzip_level] = nil #'6'
default[:varnishd][:runtime][:gzip_memlevel] = nil #'8'
default[:varnishd][:runtime][:gzip_stack_buffer] = nil #'32768'
default[:varnishd][:runtime][:gzip_tmp_space] = nil #'0'
default[:varnishd][:runtime][:gzip_window] = nil #'15'
default[:varnishd][:runtime][:http_gzip_support] = nil #'on'
default[:varnishd][:runtime][:http_max_hdr] = nil #'64'
default[:varnishd][:runtime][:http_range_support] = nil #'on'
default[:varnishd][:runtime][:http_req_hdr_len] = nil #'8192'
default[:varnishd][:runtime][:http_req_size] = nil #'32768'
default[:varnishd][:runtime][:http_resp_hdr_len] = nil #'8192'
default[:varnishd][:runtime][:http_resp_size] = nil #'32768'
default[:varnishd][:runtime][:idle_send_timeout] = nil #'60'
default[:varnishd][:runtime][:listen_address] = ':80'
default[:varnishd][:runtime][:listen_depth] = nil #'1024'
default[:varnishd][:runtime][:log_hashstring] = nil #'on'
default[:varnishd][:runtime][:log_local_address] = nil #'off'
default[:varnishd][:runtime][:lru_interval] = nil #'2'
default[:varnishd][:runtime][:max_esi_depth] = nil #'5'
default[:varnishd][:runtime][:max_restarts] = nil #'4'
default[:varnishd][:runtime][:nuke_limit] = nil #'50'
default[:varnishd][:runtime][:pcre_match_limit] = nil #'10000'
default[:varnishd][:runtime][:pcre_match_limit_recursion] = nil #'10000'
default[:varnishd][:runtime][:ping_interval] = nil #'3'
default[:varnishd][:runtime][:pipe_timeout] = nil #'60'
default[:varnishd][:runtime][:prefer_ipv6] = nil #'off'
default[:varnishd][:runtime][:queue_max] = nil #'50'
default[:varnishd][:runtime][:rush_exponent] = nil #'3'
default[:varnishd][:runtime][:saintmode_threshold] = nil #'10'
default[:varnishd][:runtime][:send_timeout] = nil #'600'
default[:varnishd][:runtime][:sess_timeout] = nil #'5'
default[:varnishd][:runtime][:sess_workspace] = nil #'65536'
default[:varnishd][:runtime][:session_linger] = nil #'50'
default[:varnishd][:runtime][:session_max] = nil #'100000'
default[:varnishd][:runtime][:shm_reclen] = nil #'255'
default[:varnishd][:runtime][:shm_workspace] = nil #'8192'
default[:varnishd][:runtime][:shortlived] = nil #'10.0'
default[:varnishd][:runtime][:syslog_cli_traffic] = nil #'on'
default[:varnishd][:runtime][:thread_pool_add_delay] = nil #'2'
default[:varnishd][:runtime][:thread_pool_add_threshold] = nil #'2'
default[:varnishd][:runtime][:thread_pool_fail_delay] = nil #'200'
default[:varnishd][:runtime][:thread_pool_max] = nil #'500'
default[:varnishd][:runtime][:thread_pool_min] = nil #'5'
default[:varnishd][:runtime][:thread_pool_purge_delay] = nil #'1000'
default[:varnishd][:runtime][:thread_pool_stack] = nil #'-1'
default[:varnishd][:runtime][:thread_pool_timeout] = nil #'300'
default[:varnishd][:runtime][:thread_pool_workspace] = nil #'65536'
default[:varnishd][:runtime][:thread_pools] = node['cpu']['total']
default[:varnishd][:runtime][:thread_stats_rate] = nil #'10'
default[:varnishd][:runtime][:user] = 'varnishd'
default[:varnishd][:runtime][:vcc_err_unref] = nil #'on'
default[:varnishd][:runtime][:vcl_dir] = nil #'/usr/local/etc/varnish'
default[:varnishd][:runtime][:vcl_trace] = nil #'off'
default[:varnishd][:runtime][:vmod_dir] = '/usr/local/lib/varnish/vmods'
default[:varnishd][:runtime][:waiter] = nil #'default'
